spring-cloud-aws-s3-parent/spring-cloud-aws-s3-cross-region-client/src/main/java/io/awspring/cloud/s3/crossregion/CrossRegionS3Client.java

```diff
@@ -99,7 +104,14 @@ private <Result> Result executeInBucketRegion(String bucket, Function<S3Client,
 			// "PermanentRedirect" means that the bucket is in different region than the
 			// defaultS3Client is configured for
 			if ("PermanentRedirect".equals(e.awsErrorDetails().errorCode())) {
-				return fn.apply(bucketCache.get(bucket));
+				e.awsErrorDetails()
+					.sdkHttpResponse()
+					.firstMatchingHeader(BUCKET_REDIRECT_HEADER)
+					.map(Region::of)
+					.map(this::getClient)
+					.ifPresent(r -> bucketCache.put(bucket, r));
+
+				return fn.apply(bucketCache.computeIfAbsent(bucket, this::discoverBucketClient));
```

Comment:
Since we can get this information from a header, I think we don't need `discoverBucketClient` any longer?

Reply:
You're right, there is no guarantee that header is there https://github.com/aws/aws-sdk-java/blob/fd409dee8ae23fb8953e0bb4dbde65536a7e0514/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/Headers.java#L226